src/audio/dsm/sof_dsm_inf.c

```diff
@@ -0,0 +1,272 @@
+}
+
+static int sof_dsm_inf_get_buffer(int32_t *buf, uint32_t frames,
+				  struct audio_stream *stream,
+				  int num_channel)
+{
+	int x, y;
+	union sof_dsm_buf input, output;
+
+	input.buf16 = (int16_t *)stream->r_ptr;
+	input.buf32 = (int32_t *)stream->r_ptr;
+	output.buf16 = (int16_t *)buf;
+	output.buf32 = (int32_t *)buf;
+
+	switch (stream->frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		for (x = 0 ; x < frames ; x++) {
+			for (y = 0 ; y < num_channel ; y++) {
+				input.buf16 =
+					(int16_t *)_wrap_buf_ptr(input.buf16,
+								 stream);
+				output.buf16[num_channel * x + y] =
+					*input.buf16;
+				input.buf16++;
+			}
+		}
+		break;
+	case SOF_IPC_FRAME_S24_4LE:
+	case SOF_IPC_FRAME_S32_LE:
+		for (x = 0 ; x < frames ; x++) {
+			for (y = 0 ; y < num_channel ; y++) {
+				input.buf32 =
+					(int32_t *)_wrap_buf_ptr(input.buf32,
+								 stream);
+				output.buf32[num_channel * x + y] =
+					*input.buf32;
+				input.buf32++;
+			}
+		}
+		break;
+	default:
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int sof_dsm_inf_put_buffer(int32_t *buf, uint32_t frames,
+				  struct audio_stream *stream, int num_channel,
+				  int num_slot, char mask, struct comp_dev *dev)
+{
+	int x, y, offset;
+	char value;
+	union sof_dsm_buf input, output;
+	const char pow_2[8] = {1, 2, 4, 8, 16, 32, 64, 128};
+
+	input.buf16 = (int16_t *)buf;
+	input.buf32 = (int32_t *)buf;
+	output.buf16 = (int16_t *)stream->w_ptr;
+	output.buf32 = (int32_t *)stream->w_ptr;
+
+	switch (stream->frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		for (x = 0 ; x < frames ; x++) {
+			for (y = 0 ; y < num_slot ; y++) {
+				if (y == 0)
+					offset = 0;
+				output.buf16 =
+					(int16_t *)_wrap_buf_ptr(output.buf16,
+								 stream);
+				value = pow_2[y] & mask;
+				*output.buf16 =
+					value > 0 ? input.buf16[num_channel * x
+					+ (offset++)] : 0;
+				output.buf16++;
+			}
+		}
+		break;
+	case SOF_IPC_FRAME_S24_4LE:
+	case SOF_IPC_FRAME_S32_LE:
+		for (x = 0 ; x < frames ; x++) {
+			for (y = 0 ; y < num_slot ; y++) {
+				if (y == 0)
+					offset = 0;
+				output.buf32 =
+					(int32_t *)_wrap_buf_ptr(output.buf32,
+								 stream);
+				value = pow_2[y] & mask;
+				*output.buf32 =
+					value > 0 ? input.buf32[num_channel * x
+					+ (offset++)] : 0;
+				output.buf32++;
+			}
+		}
+		break;
+	default:
+		return -EINVAL;
+	}
+	return 0;
+}
+
+int sof_dsm_inf_ff_copy(struct comp_dev *dev, uint32_t frames,
+			struct comp_buffer *source,
+			struct comp_buffer *sink, int8_t *chan_map,
+			struct sof_dsm_struct_t *hsof_dsm)
+{
+	int ret;
+
+	if (!hsof_dsm->dsm_init)
+		return -EINVAL;
+
+	hsof_dsm->seq_sof_in++;
+
+	if (frames == 0) {
+		comp_dbg(dev, "[DSM] FF frame size zero warning.");
+		return 0;
+	}
+
+	if (frames > SOF_FF_BUF_DB_SZ) {
+		comp_err(dev, "[DSM] FF number of frame error  : %d", frames);
+		return -EINVAL;
+	}
+
+	ret = sof_dsm_inf_get_buffer(hsof_dsm->buf.sof_a_frame_in,
+				     frames, &source->stream, 2);
+	if (ret != 0)
```

Comment:
checkpatch wants ```if (!ret)``` IIRC.

Reply:
Thanks. It is strange why my checkpatch(./script/checkpantch) don't complain this. Shall fix this.